al-khaser/Anti Debug/Interrupt_0x2d.cpp

```diff
@@ -4,32 +4,32 @@
 The Interrupt_0x2d function will check to see if a debugger is attached to the current process. It does this by setting up
 SEH and using the Int 2D instruction which will only cause an exception if there is no debugger. Also when used in OllyDBG
 it will skip a byte in the disassembly which could be used to detect the debugger.
-Atm, only x86 version is available in VC++, in x64 I couln't find any __int2d in msdn
+
+Vectored Exception Handling is used here because SEH is an anti-debug trick in itself.
 */
 
-BOOL Interrupt_0x2d()
-{
-	CHAR IsBeingDbg = 0;
+extern "C" void __int2d();
 
-	__try
-	{
-		__asm
-		{
-			xor eax, eax
-			int 0x2d
-			inc eax
-			je DebuggerFound
-			jmp Exit
-		DebuggerFound :
-			mov IsBeingDbg, 1
-		Exit:
-		}
+static BOOL SwallowedException = TRUE;
 
-	}
-	__except (GetExceptionCode() == EXCEPTION_BREAKPOINT ? EXCEPTION_EXECUTE_HANDLER : EXCEPTION_CONTINUE_SEARCH)
+static LONG CALLBACK VectoredHandler(
+	_In_ PEXCEPTION_POINTERS ExceptionInfo
+)
+{
+	SwallowedException = FALSE;
+	if (ExceptionInfo->ExceptionRecord->ExceptionCode == EXCEPTION_BREAKPOINT)
 	{
-		IsBeingDbg = FALSE;
+		//The Int 2D instruction already increased EIP/RIP so we don't do that (although it wouldnt hurt).
+		return EXCEPTION_CONTINUE_EXECUTION;
 	}
+	return EXCEPTION_CONTINUE_SEARCH;
+}
 
-	return IsBeingDbg;
+BOOL Interrupt_0x2d()
+{
+	PVOID Handle = AddVectoredExceptionHandler(1, VectoredHandler);
+	SwallowedException = TRUE;
+	__int2d();
+	RemoveVectoredExceptionHandler(Handle);
+	return SwallowedException;

```

Comment:
Actually SEH should work, but I think that because of the way seh works in
x64 it will not catch the int2d opcode.
On Fri, 1 Jul 2016 at 10:54, Noteworthy <notifications@github.com> wrote:

> In al-khaser/Anti Debug/Interrupt_0x2d.cpp
> <https://github.com/LordNoteworthy/al-khaser/pull/50#discussion_r69269039>
> :
>
> >
> > -	return IsBeingDbg;
> > +BOOL Interrupt_0x2d()
> > +{
> > +	PVOID Handle = AddVectoredExceptionHandler(1, VectoredHandler);
> > +	SwallowedException = TRUE;
> > +	__int2d();
> > +	RemoveVectoredExceptionHandler(Handle);
> > +	return SwallowedException;
>
> Thanks for information (The VEH). Good that you added 64-bits support. The
> same applies for Interrupt_3().
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/LordNoteworthy/al-khaser/pull/50/files/e676155ffe849064e4402230da24b23978bc09eb#r69269039>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACWCmbIt7eqEg7Z7ox6Ep_dqQSHKOguDks5qRNXIgaJpZM4I_rAC>
> .
>